service/collector/vmss.go

```diff
@@ -0,0 +1,234 @@
+	{
+		for _, cr := range crs {
+			subscriptionCredentials := fmt.Sprintf("%s-%s", key.CredentialName(cr), key.CredentialNamespace(cr))
+			crsBySubscription[subscriptionCredentials] = cr
+		}
+	}
+
+	ctx := context.Background()
+
+	// We track VMSS metrics for each client labeled by ClientID.
+	// That way we prevent duplicated metrics.
+	for _, cr := range crsBySubscription {
+		azureConfig, azureClients, err := u.getAzureClients(cr)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		var reads float64
+		{
+			vmssResponse, err := azureClients.VirtualMachineScaleSetsClient.ListAll(ctx)
+			if err != nil {
+				u.logger.Log("level", "warning", "message", "an error occurred during the scraping of current compute resource VMSS information", "stack", fmt.Sprintf("%v", err))
+			} else {
+				reads, err = strconv.ParseFloat(vmssResponse.Response().Header.Get(remainingReadsHeaderName), 64)
+				if err != nil {
+					reads = 0
+				}
+
+				ch <- prometheus.MustNewConstMetric(
+					VMSSReadsDesc,
+					prometheus.GaugeValue,
+					reads,
+					azureClients.VirtualMachineScaleSetsClient.SubscriptionID,
+					azureConfig.ClientID,
+				)
+			}
+		}
+
+		// Remaining write requests can be retrieved sending a write request.
+		// We create a resource group with a random name, so we can delete it.
+		var writes float64
+		{
+			resourceGroupName := randStringBytes(16)
+			_, err := azureClients.GroupsClient.CreateOrUpdate(
+				ctx,
+				resourceGroupName,
+				resources.Group{
+					Location: to.StringPtr(u.location),
+				})
+
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			deleteResponse, err := azureClients.GroupsClient.Delete(ctx, resourceGroupName)
+
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			writes, err = strconv.ParseFloat(deleteResponse.Response().Header.Get(remainingWritesHeaderName), 64)
+			if err != nil {
+				writes = 0
+			}
+
+			ch <- prometheus.MustNewConstMetric(
+				VMSSWritesDesc,
+				prometheus.GaugeValue,
+				writes,
+				azureClients.GroupsClient.SubscriptionID,
+				azureConfig.ClientID,
+			)
+		}
+	}
+
+	return nil
+}
+
+func (u *VMSS) Describe(ch chan<- *prometheus.Desc) error {
+	ch <- VMSSReadsDesc
+	return nil
+}
+
+func (u *VMSS) getAzureClients(cr providerv1alpha1.AzureConfig) (*client.AzureClientSetConfig, *client.AzureClientSet, error) {
+	config, err := credential.GetAzureConfig(u.k8sClient, key.CredentialName(cr), key.CredentialNamespace(cr))
+	if err != nil {
+		return nil, nil, microerror.Mask(err)
+	}
+	config.EnvironmentName = u.environmentName
+
+	azureClients, err := client.NewAzureClientSet(*config)
+	if err != nil {
+		return nil, nil, microerror.Mask(err)
+	}
+
+	return config, azureClients, nil
+}
+
+const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
+
+func randStringBytes(n int) string {
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = letterBytes[rand.Intn(len(letterBytes))]
+	}
+
+	return string(b)
+}
```

Comment:
should I use UUID's instead of generating my own random alphanumeric string? What do you think?

Reply:
You could use UUIDs, yes. That will make it 100% unique

Reply:
I figure random string is fine. UUID could be easily confusing when seen because Microsoft uses GUIDs for everything.